### PR TITLE
Optimize writing single segment sequences

### PR DESCRIPTION
**Pull request by JamesNK**
Avoid foreach over `ReadOnlySequence<byte>` when it is a single segment.

### CODE REVIEW

**Comment by davidfowl** [comment]
What kinda gains are you seeing?

**Comment by JamesNK** [comment]
I didn't microbenchmark. It showed up in benchmark and seems like a simple thing to improve.

image

File: src/Shared/ServerInfrastructure/BufferExtensions.cs [hunk 1]
```diff
@@ -26,6 +26,27 @@ public static ReadOnlySpan<byte> ToSpan(in this ReadOnlySequence<byte> buffer)
             return buffer.ToArray();
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void CopyTo(in this ReadOnlySequence<byte> buffer, PipeWriter pipeWriter)
```
Comment (davidfowl) [inline src/Shared/ServerInfrastructure/BufferExtensions.cs:30]: Maybe we should add this to ReadOnlySequence the target would be an IBufferWriter

Comment (JamesNK) [reply src/Shared/ServerInfrastructure/BufferExtensions.cs:30]: Add it to runtime? Could do. It has this today:

```cs
    public static class BuffersExtensions
    {
        public static void Write<T>(this IBufferWriter<T> writer, ReadOnlySpan<T> value)
        {
        }
    }
```

Comment (JamesNK) [reply src/Shared/ServerInfrastructure/BufferExtensions.cs:30]: https://github.com/dotnet/runtime/issues/25443

**Review by Tratcher — APPROVED** [review]

**Comment by davidfowl** [comment]
Put this in the RC1 branch

**Pull request merged**